File: 05-Factory-Method/src/go/transport.go
package main

import "fmt"

// Product interface
type Transport interface {
	Deliver()
	Name() string
}

type Truck struct{}
func (Truck) Deliver() { fmt.Println("Delivering by land in a truck.") }
func (Truck) Name() string { return "Truck" }

type Ship struct{}
func (Ship) Deliver() { fmt.Println("Delivering by sea in a container ship.") }
func (Ship) Name() string { return "Ship" }

type Airplane struct{}
func (Airplane) Deliver() { fmt.Println("Delivering by air in a cargo airplane.") }
func (Airplane) Name() string { return "Airplane" }

File: 05-Factory-Method/src/rust/transport.rs
// Product trait
pub trait Transport {
    fn deliver(&self);
    fn name(&self) -> &str;
}

pub struct Truck;
impl Transport for Truck {
    fn deliver(&self) { println!("Delivering by land in a truck."); }
    fn name(&self) -> &str { "Truck" }
}

pub struct Ship;
impl Transport for Ship {
    fn deliver(&self) { println!("Delivering by sea in a container ship."); }
    fn name(&self) -> &str { "Ship" }
}

pub struct Airplane;
impl Transport for Airplane {
    fn deliver(&self) { println!("Delivering by air in a cargo airplane."); }
    fn name(&self) -> &str { "Airplane" }
}
